src/rust/src/x509.rs

```diff
@@ -43,6 +43,454 @@ lazy_static::lazy_static! {
+        if encoding == encoding_class.getattr("DER")? {
+            Ok(pyo3::types::PyBytes::new(py, &result))
+        } else if encoding == encoding_class.getattr("PEM")? {
+            let pem = pem::encode_config(
+                &pem::Pem {
+                    tag: "CERTIFICATE".to_string(),
+                    contents: result,
+                },
+                pem::EncodeConfig {
+                    line_ending: pem::LineEnding::LF,
+                },
+            )
+            .into_bytes();
+            Ok(pyo3::types::PyBytes::new(py, &pem))
+        } else {
+            Err(pyo3::exceptions::PyTypeError::new_err(
+                "encoding must be an item from the Encoding enum",
+            ))
+        }
+    }
+
+    #[getter]
+    fn serial_number<'p>(&self, py: pyo3::Python<'p>) -> Result<&'p pyo3::PyAny, PyAsn1Error> {
+        Ok(big_asn1_uint_to_py(py, self.raw.borrow_value().tbs_cert.serial)?)
+    }
+
+    #[getter]
+    fn version<'p>(&self, py: pyo3::Python<'p>) -> Result<&'p pyo3::PyAny, PyAsn1Error> {
+        let x509_module = py.import("cryptography.x509")?;
+        let version = self.raw.borrow_value().tbs_cert.version;
+        if version == 0 {
+            Ok(x509_module.getattr("Version")?.get_item("v1")?)
+        } else {
+            // in the load function we prohibit anything but 0 or 2
+            assert!(version == 2);
+            Ok(x509_module.getattr("Version")?.get_item("v3")?)
+        }
+    }
+
+    #[getter]
+    fn issuer<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        parse_name(py, &self.raw.borrow_value().tbs_cert.issuer)
+    }
+
+    #[getter]
+    fn subject<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        parse_name(py, &self.raw.borrow_value().tbs_cert.subject)
+    }
+
+    #[getter]
+    fn tbs_certificate_bytes<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+    ) -> Result<&'p pyo3::types::PyBytes, PyAsn1Error> {
+        let result = asn1::write_single(&self.raw.borrow_value().tbs_cert);
+        Ok(pyo3::types::PyBytes::new(py, &result))
+    }
+
+    #[getter]
+    fn signature<'p>(&self, py: pyo3::Python<'p>) -> Result<&'p pyo3::types::PyBytes, PyAsn1Error> {
+        Ok(pyo3::types::PyBytes::new(
+            py,
+            self.raw.borrow_value().signature.as_bytes(),
+        ))
+    }
+
+    #[getter]
+    fn not_valid_before<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let time = match &self.raw.borrow_value().tbs_cert.validity.not_before {
+            Asn1Time::UtcTime(data) => data.as_chrono(),
+            Asn1Time::GeneralizedTime(data) => data.as_chrono(),
+        };
+        chrono_to_py(py, time)
+    }
+
+    #[getter]
+    fn not_valid_after<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let time = match &self.raw.borrow_value().tbs_cert.validity.not_after {
```

Comment:
Perhaps an `as_chrono` method on `Time` with this match, since we do it in a few places?